### PR TITLE
feat(database): init normalisation tool

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

### CODE REVIEW

File: database/migrations/mikoyan/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        match normalize_user(&user) {
+            Ok(normalized_user) => {
+                // _id exists, because `normalize_user` returns an error if it does not
+                let id = user.get_object_id("_id").unwrap();
+                let filter = doc! {"_id": id};
+                let update_options = UpdateOptions::builder()
+                    .array_filters(vec![doc! {
+                        "el": {"$exists": true},
+                    }])
+                    .build();
```
Comment (ojeytonwilliams) [inline database/migrations/mikoyan/src/main.rs:136]: What does this do?

**Review by ojeytonwilliams — COMMENTED** [review]
Hey, @ShaunSHamilton, could you add some tests for this? The logic for updating completed challenges and so on is pretty complicated, so it would be good to see a few tests of the transformation.

File: database/migrations/mikoyan/src/main.rs (hunk 1 above)
Comment (ShaunSHamilton) [reply database/migrations/mikoyan/src/main.rs:136]: Docs: https://www.mongodb.com/docs/manual/reference/operator/update/positional-filtered/

`normalize.rs` has this:

```rust
"completedChallenges.$[el].__cachedRelations": "",
"completedChallenges.$[el].__data": "",
"completedChallenges.$[el].__dataSource": "",
"completedChallenges.$[el].__persisted": "",
"completedChallenges.$[el].__strict": "",
```

So this defines `el`, and prevents errors by only performing the `$[el]` update when it exists.

**Comment by ShaunSHamilton** [comment]
> could you add some tests for this? The logic for updating completed challenges and so on is pretty complicated

I think the logic is very simple, but could add tests. I do not see myself having time anytime soon, because it is quite involved writing tests for Mongodb operations (the fixture data is annoying).

**Comment by ojeytonwilliams** [comment]
We do need to be confident in that scripts that modify user data, and some tests would really help with that.  I don't really understand Rust, but something along the lines of 

```rust
let unnormalized_user = doc! { blah }
let normalized_user = doc! { blah blah}
let result = normalized_user(&unnormalized_user)
// something to assert deep equality
```

should be possible, shouldn't it?

**Comment by naomi-lgbt** [comment]
can I block this on the principle that rust doesn't belong in the codebase? 🙃

**Comment by ShaunSHamilton** [comment]
> can I block this on the principle that rust doesn't belong in the codebase? 🙃

Once TypeScript becomes correct and statically typed, I will rewrite this in it.

**Comment by ShaunSHamilton** [comment]
Closing as this is not the direction we plan to go. A new PR will be made to handle _"normalization"_-type tasks

**Pull request closed without merging**